applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp

```diff
@@ -468,5 +479,210 @@ void FemDem3DLargeDisplacementElement::CalculateAndAddInternalForcesVector(
     noalias(rRightHandSideVector) -= IntegrationWeight * prod(trans(rB), rStressVector);
 }
 
+// Methods to compute the tangent tensor by numerical derivation
+void FemDem3DLargeDisplacementElement::CalculateTangentTensor(
+	Matrix& rTangentTensor,
+	const Vector& rStrainVectorGP,
+	const Vector& rStressVectorGP,
+    const Matrix& rDeformationGradientGP,
+	const Matrix& rElasticMatrix
+	)
+{
+	const double number_components = rStrainVectorGP.size();
+	rTangentTensor.resize(number_components, number_components);
+	Vector perturbed_stress, perturbed_strain;
+	perturbed_strain.resize(number_components);
+	perturbed_stress.resize(number_components);
+    Matrix perturbed_deformation_gradient;
+    const double size_1 = rDeformationGradientGP.size1();
+    const double size_2 = rDeformationGradientGP.size2();
+	
+	for (unsigned int i_component = 0; i_component < size_1; i_component++) {
+	    for (unsigned int j_component = i_component; j_component < size_2; j_component++) {
+            double perturbation;
+            const int component_voigt_index = this->CalculateVoigtIndex(number_components, i_component, j_component);
+            this->CalculatePerturbation(rStrainVectorGP, perturbation, component_voigt_index);
+            this->PerturbateDeformationGradient(perturbed_deformation_gradient, rDeformationGradientGP, perturbation, i_component, j_component);
+            this->CalculateGreenLagrangeStrainVector(perturbed_strain, perturbed_deformation_gradient);
+            this->IntegratePerturbedStrain(perturbed_stress, perturbed_strain, rElasticMatrix);
+            const Vector& r_delta_stress = perturbed_stress - rStressVectorGP;
+            this->AssignComponentsToTangentTensor(rTangentTensor, r_delta_stress, perturbation, component_voigt_index);
+        }
+	}
+}
+
+void FemDem3DLargeDisplacementElement::PerturbateDeformationGradient(
+    Matrix& rPerturbedDeformationGradient,
+    const Matrix& rDeformationGradientGP,
+    const double Perturbation,
+    const int ComponentI,
+    const int ComponentJ
+)
+{
+    rPerturbedDeformationGradient = rDeformationGradientGP;
+    if (ComponentI == ComponentJ) {
+        rPerturbedDeformationGradient(ComponentI, ComponentJ) += Perturbation;
+    } else {
+        rPerturbedDeformationGradient(ComponentI, ComponentJ) += 0.5 * Perturbation;
+        rPerturbedDeformationGradient(ComponentJ, ComponentI) += 0.5 * Perturbation;
+    }
+}
+
+void FemDem3DLargeDisplacementElement::IntegratePerturbedDeformationGradient(
+    Vector& rPerturberStressVector,
+    const Vector& rPerturbedStrainVector,
+    const Matrix& rElasticMatrix,
+    const Matrix& rPerturbedDeformationGradient
+    )
+{
+}
+
+
+int FemDem3DLargeDisplacementElement::CalculateVoigtIndex(
+    const SizeType VoigtSize,
+    const int ComponentI,
+    const int ComponentJ
+    )
+{
+    if (VoigtSize == 6) {

```

Comment:
I think this is in MathUtils

Reply:
I see it in the tangent tensor calculations :)